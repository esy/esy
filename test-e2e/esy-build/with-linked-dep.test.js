// @flow

const path = require('path');
const fs = require('fs');
const {promisify} = require('util');
const open = promisify(fs.open);
const close = promisify(fs.close);

const helpers = require('../test/helpers');
const {test, isWindows, isMacos, isLinux} = helpers;

function makeFixture(p, buildDep) {
  return [
    helpers.packageJson({
      name: 'with-linked-dep-_build',
      version: '1.0.0',
      esy: {
        build: 'true',
      },
      dependencies: {
        dep: '*',
      },
      resolutions: {
        dep: 'link:./dep',
      },
    }),
    helpers.dir(
      'dep',
      helpers.packageJson({
        name: 'dep',
        version: '1.0.0',
        esy: buildDep,
      }),
      helpers.dummyExecutable('dep'),
    ),
  ];
}

describe('Build with a linked dep', () => {
  describe('out of source build', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            build: [
              'cp #{self.root / self.name}.js #{self.target_dir / self.name}.js',
              helpers.buildCommand(p, '#{self.target_dir / self.name}.js'),
            ],
            install: [
              `cp #{self.target_dir / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.target_dir / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it('package "dep" should be visible in all envs', withProject(async p => {
      {
        const {stdout} = await p.esy('dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('b dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('x dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }
    }));

    test.enableIf(isMacos || isLinux)(
      'macos || linux: build-env snapshot',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan')).stdout).id;
        const depid = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const {stdout} = await p.esy('build-env --build-concurrency 16');
        expect(p.normalizePathsForSnapshot(stdout, {id, depid})).toMatchSnapshot();
      }),
    );

    test.enableIf(isMacos || isLinux)(
      'macos || linux: build-env dep snapshot',
      withProject(async function(p) {
        const id = JSON.parse((await p.esy('build-plan -p dep')).stdout).id;
        const {stdout} = await p.esy('build-env -p dep --build-concurrency 16');
        expect(p.normalizePathsForSnapshot(stdout, {id})).toMatchSnapshot();
      }),
    );
  });

  describe('in source build', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            buildsInSource: true,
            build: [helpers.buildCommand(p, '#{self.root / self.name}.js')],
            install: [
              `cp #{self.root / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.root / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it('package "dep" should be visible in all envs', withProject(async p => {
      {
        const {stdout} = await p.esy('dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('b dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('x dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }
    }));
  });

  describe('_build build', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            buildsInSource: '_build',
            build: [
              "mkdir -p #{self.root / '_build'}",
              "cp #{self.root / self.name}.js #{self.root / '_build' / self.name}.js",
              helpers.buildCommand(p, "#{self.root / '_build' / self.name}.js"),
            ],
            install: [
              `cp #{self.root / '_build' / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.root / '_build' / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it('package "dep" should be visible in all envs', withProject(async p => {
      {
        const {stdout} = await p.esy('dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('b dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('x dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }
    }));
  });

  describe('out of source build with "buildDev" in dep', () => {
    function withProject(assertions) {
      return async () => {
        const p = await helpers.createTestSandbox();
        await p.fixture(
          ...makeFixture(p, {
            build: [
              'cp #{self.root / self.name}.js #{self.target_dir / self.name}.js',
              helpers.buildCommand(p, '#{self.target_dir / self.name}.js'),
            ],
            // set it to false so we fail if it's executed
            buildDev: 'false',
            install: [
              `cp #{self.target_dir / self.name}.cmd #{self.bin / self.name}.cmd`,
              `cp #{self.target_dir / self.name}.js #{self.bin / self.name}.js`,
            ],
          }),
        );
        await p.esy('install');
        await p.esy('build');
        await assertions(p);
      };
    }

    it('package "dep" should be visible in all envs', withProject(async p => {
      {
        const {stdout} = await p.esy('dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('b dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }

      {
        const {stdout} = await p.esy('x dep.cmd');
        expect(stdout.trim()).toEqual('__dep__');
      }
    }));
  });

});
