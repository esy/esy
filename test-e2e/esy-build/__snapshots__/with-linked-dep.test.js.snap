// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build with a linked dep out of source build macos || linux: build-env dep snapshot 1`] = `
"# Build environment
# package:            dep@link:dep
# depspec:            dependencies(self)
# mode:               build
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%projectPath%/_esy/default/store/i/%id%/etc"
export cur__share="%projectPath%/_esy/default/store/i/%id%/share"
export cur__toplevel="%projectPath%/_esy/default/store/i/%id%/toplevel"
export cur__stublibs="%projectPath%/_esy/default/store/i/%id%/stublibs"
export cur__doc="%projectPath%/_esy/default/store/i/%id%/doc"
export cur__man="%projectPath%/_esy/default/store/i/%id%/man"
export cur__lib="%projectPath%/_esy/default/store/i/%id%/lib"
export cur__sbin="%projectPath%/_esy/default/store/i/%id%/sbin"
export cur__bin="%projectPath%/_esy/default/store/i/%id%/bin"
export cur__install="%projectPath%/_esy/default/store/i/%id%"
export cur__target_dir="%projectPath%/_esy/default/store/b/%id%"
export cur__original_root="%projectPath%/dep"
export cur__root="%projectPath%/dep"
export cur__dev="false"
export cur__version="1.0.0"
export cur__name="dep"
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%projectPath%/_esy/default/store/i/%id%/lib"
export DUNE_BUILD_DIR="%projectPath%/_esy/default/store/b/%id%"
export DUNE_STORE_ORIG_SOURCE_DIR="true"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;

exports[`Build with a linked dep out of source build macos || linux: build-env snapshot 1`] = `
"# Build environment
# package:            with-linked-dep-_build@link-dev:./package.json
# depspec:            dependencies(self)+devDependencies(self)
# mode:               buildDev
# envspec:            
# buildIsInProgress:  true
# includeBuildEnv:    true
# includeCurrentEnv:  false
# includeNpmBin:      false


#
# Built-in
#
export cur__jobs="8"
export cur__etc="%projectPath%/_esy/default/store/i/%id%/etc"
export cur__share="%projectPath%/_esy/default/store/i/%id%/share"
export cur__toplevel="%projectPath%/_esy/default/store/i/%id%/toplevel"
export cur__stublibs="%projectPath%/_esy/default/store/i/%id%/stublibs"
export cur__doc="%projectPath%/_esy/default/store/i/%id%/doc"
export cur__man="%projectPath%/_esy/default/store/i/%id%/man"
export cur__lib="%projectPath%/_esy/default/store/i/%id%/lib"
export cur__sbin="%projectPath%/_esy/default/store/i/%id%/sbin"
export cur__bin="%projectPath%/_esy/default/store/i/%id%/bin"
export cur__install="%projectPath%/_esy/default/store/i/%id%"
export cur__target_dir="%projectPath%/_esy/default/store/b/%id%"
export cur__original_root="%projectPath%"
export cur__root="%projectPath%"
export cur__dev="true"
export cur__version="1.0.0"
export cur__name="with-linked-dep-_build"

#
# dep@link:dep
#
export CAML_LD_LIBRARY_PATH="%projectPath%/_esy/default/store/i/%depid%/stublibs:%projectPath%/_esy/default/store/i/%depid%/lib/stublibs:$CAML_LD_LIBRARY_PATH"
export OCAMLPATH="%projectPath%/_esy/default/store/i/%depid%/lib:$OCAMLPATH"
export MAN_PATH="%projectPath%/_esy/default/store/i/%depid%/man:$MAN_PATH"
export PATH="%projectPath%/_esy/default/store/i/%depid%/bin:$PATH"

#
# Built-in
#
export OCAMLFIND_LDCONF="ignore"
export OCAMLFIND_DESTDIR="%projectPath%/_esy/default/store/i/%id%/lib"
export DUNE_BUILD_DIR="%projectPath%/_esy/default/store/b/%id%"
export DUNE_STORE_ORIG_SOURCE_DIR="true"
export SHELL="env -i /bin/bash --norc --noprofile"
export PATH="$PATH:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin""
`;
